File: doubly_linked_list/doubly_linked_list.go
package doublylinkedlist

import "fmt"

type Node struct {
	Value      string
	Next, Prev *Node
}

func (node Node) String() string {
	if node.Next != nil {
		return fmt.Sprintf("[Node: %v. Next: %v]", node.Value, node.Next.Value)
	} else {
		return fmt.Sprintf("[Node: %v]", node.Value)
	}
}

type DoublyLinkedList struct {
	head *Node
}

func (list *DoublyLinkedList) GetHead() *Node {
	return list.head
}

func (list *DoublyLinkedList) InsertHead(value string) (node *Node) {
	node = &Node{value, nil, nil}
	if list.head == nil {
		list.head = node
		list.head.Prev = nil
		list.head.Next = nil
	} else {
		node.Next = list.head
		node.Prev = nil
		list.head = node
	}
	return node
}

func (list *DoublyLinkedList) Insert(value string) *Node {
	if list.head == nil {
		return list.InsertHead(value)
	}
	last := list.head
	for last.Next != nil {
		last = last.Next
	}
	node := &Node{value, nil, last}
	last.Next = node

	return node
}

func (list *DoublyLinkedList) InsertAfter(prevNode *Node, value string) (newItem *Node) {
	if prevNode == nil {
		return nil
	}
	newItem = &Node{value, prevNode.Next, prevNode}
	prevNode.Next = newItem
	return newItem
}

func (list *DoublyLinkedList) InsertAt(position uint, value string) (node *Node) {
	prevItem := list.FindByIndex(int(position) - 1)
	if prevItem == nil {
		return nil
	}
	node = &Node{value, prevItem.Next, prevItem}
	prevItem.Next = node

	return node
}

func (list *DoublyLinkedList) Find(value string) *Node {
	current := list.head
	for current != nil {
		if current.Value == value {
			return current
		}
		current = current.Next
	}
	return nil
}

func (list *DoublyLinkedList) FindByIndex(index int) (item *Node) {
	if index < 0 {
		return nil
	}

	i := 0
	temp := list.head
	for i <= index || item == nil || temp != nil {
		if i == index {
			item = temp
		}
		temp = temp.Next
		i++
	}
	return item
}

func (list *DoublyLinkedList) DeleteHead() {
	if list.head == nil {
		return
	}
	newHead := list.head.Next
	list.head.Next = nil
	list.head = newHead
}

func (list *DoublyLinkedList) DeleteTail() {
	if list.head == nil {
		return
	}
	if list.head.Next == nil {
		list.head = nil
	}
	secondToLast := list.head
	for secondToLast.Next.Next != nil {
		secondToLast = secondToLast.Next
	}
	secondToLast.Next = nil
}

func (list *DoublyLinkedList) DeleteAt(index uint) {
	if index == 0 {
		list.DeleteHead()
	}
	prevItem := list.FindByIndex(int(index) - 1)
	if prevItem == nil || prevItem.Next == nil {
		return
	}
	prevItem.Next = prevItem.Next.Next
}

func (list *DoublyLinkedList) BubbleSort() {
	current := list.head
	for current != nil {
		index := current.Next
		for index != nil {
			if current.Value > index.Value {
				temp := current.Value
				current.Value = index.Value
				index.Value = temp
			}
			index = index.Next
		}
		current = current.Next
	}
}

func (list DoublyLinkedList) String() string {
	result := fmt.Sprintf("[%v", list.head.Value)
	temp := list.head.Next
	for temp != nil {
		result += fmt.Sprintf(" --> ")
		result += fmt.Sprintf("%v", temp.Value)
		temp = temp.Next
	}
	result += fmt.Sprintf("]")
	return result
}

File: doubly_linked_list/doubly_linked_list.js
export class ValueNode {
	/** @type {string} */
	#value

	get value() {
		return this.#value
	}

	/** @type {ValueNode} */
	#next = null

	get next() {
		return this.#next
	}

	set next(node) {
		if (this.#next === null) {
			this.#next = node
		}
	}

	/** @type {ValueNode} */
	#prev = null

	get prev() {
		return this.#prev
	}

	set prev(node) {
		if (this.#prev === null) {
			this.#prev = node
		}
	}

	/**
	 * @param {string} value
	 */
	constructor(value) {
		this.#value = value
	}

	getHead() {
		const recursiveHead = (node) => node.prev !== null ? recursiveHead(node.prev) : node
		return recursiveHead(this)
	}

	*[Symbol.iterator]() {
		let node = this

		while (node !== null) {
			yield node
			node = node.next
		}
	}
}

export class DoublyLinkedList {
	/** @type {ValueNode} */
	#head = null

	/** @type {ValueNode} */
	#tail = null

	/** @param {ValueNode} node */
	constructor(node = null) {
		if (!!node) {
			this.#head = this.#tail = node
		}
	}

	/** @param {ValueNode} node */
	append(node) {
		if (!this.#head) {
			this.#head = node
		} else {
			this.#tail.next = node
		}
		this.#tail = node
	}

	*[Symbol.iterator]() {
		let node = this.#head

		while (node !== null) {
			yield node
			node = node.next
		}
	}
}
